requestmanager/client.go

```diff
@@ -0,0 +1,359 @@
+	p peer.ID,
+	root ipld.Link,
+	selectorNode ipld.Node,
+	extensions ...graphsync.ExtensionData) (<-chan graphsync.ResponseProgress, <-chan error) {
+	if _, err := selector.ParseSelector(selectorNode); err != nil {
+		return rm.singleErrorResponse(fmt.Errorf("invalid selector spec"))
+	}
+
+	inProgressRequestChan := make(chan inProgressRequest)
+
+	rm.send(&newRequestMessage{p, root, selectorNode, extensions, inProgressRequestChan}, ctx.Done())
+	var receivedInProgressRequest inProgressRequest
+	select {
+	case <-rm.ctx.Done():
+		return rm.emptyResponse()
+	case receivedInProgressRequest = <-inProgressRequestChan:
+	}
+
+	// If the connection to the peer is disconnected, fire an error
+	unsub := rm.listenForDisconnect(p, func(neterr error) {
+		rm.networkErrorListeners.NotifyNetworkErrorListeners(p, receivedInProgressRequest.request, neterr)
+	})
+
+	return rm.rc.collectResponses(ctx,
+		receivedInProgressRequest.incoming,
+		receivedInProgressRequest.incomingError,
+		func() {
+			rm.cancelRequestAndClose(receivedInProgressRequest.requestID,
+				receivedInProgressRequest.incoming,
+				receivedInProgressRequest.incomingError)
+		},
+		// Once the request has completed, stop listening for disconnect events
+		unsub,
+	)
+}
+
+// Dispatch the Disconnect event to subscribers
+func disconnectDispatcher(p pubsub.Event, subscriberFn pubsub.SubscriberFn) error {
+	listener := subscriberFn.(func(peer.ID))
+	listener(p.(peer.ID))
+	return nil
+}
+
+// Listen for the Disconnect event for the given peer
+func (rm *RequestManager) listenForDisconnect(p peer.ID, onDisconnect func(neterr error)) func() {
+	// Subscribe to Disconnect notifications
+	return rm.disconnectNotif.Subscribe(func(evtPeer peer.ID) {
+		// If the peer is the one we're interested in, call the listener
+		if evtPeer == p {
+			onDisconnect(fmt.Errorf("disconnected from peer %s", p))
+		}
+	})
+}
+
+// Disconnected is called when a peer disconnects
+func (rm *RequestManager) Disconnected(p peer.ID) {
+	// Notify any listeners that a peer has disconnected
+	_ = rm.disconnectNotif.Publish(p)
+}
+
+func (rm *RequestManager) emptyResponse() (chan graphsync.ResponseProgress, chan error) {
+	ch := make(chan graphsync.ResponseProgress)
+	close(ch)
+	errCh := make(chan error)
+	close(errCh)
+	return ch, errCh
+}
+
+func (rm *RequestManager) singleErrorResponse(err error) (chan graphsync.ResponseProgress, chan error) {
+	ch := make(chan graphsync.ResponseProgress)
+	close(ch)
+	errCh := make(chan error, 1)
+	errCh <- err
+	close(errCh)
+	return ch, errCh
+}
+
+func (rm *RequestManager) cancelRequestAndClose(requestID graphsync.RequestID,
+	incomingResponses chan graphsync.ResponseProgress,
+	incomingErrors chan error) {
+	cancelMessageChannel := rm.messages
+	for cancelMessageChannel != nil || incomingResponses != nil || incomingErrors != nil {
+		select {
+		case cancelMessageChannel <- &cancelRequestMessage{requestID, false, nil, nil}:
+			cancelMessageChannel = nil
```

Comment:
should this close the channel as well? a comment might be useful since this pattern looks weird otherwise

Reply:
this if you look a couple lines above, what we're doing here is using cancelMessageChannel = nil to send exactly once -- but the actual channel is the main message channel for the request manager.

I agree, this is overcomplicated. I think it's probably ok to do the channel send once before the loop, but for this PR I'd rather not change that functionality.